hal/armv8m/interrupts.c

```diff
@@ -0,0 +1,112 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * Interrupt handling - NVIC (Nested Vectored Interrupt Controller)
+ *
+ * Copyright 2017, 2019, 2020 Phoenix Systems
+ * Author: Aleksander Kaminski, Jan Sikorski, Hubert Buczynski, Damian Loewnau
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <hal/hal.h>
+
+
+typedef struct {
+	void *data;
+	int (*isr)(unsigned int, void *);
+} intr_handler_t;
+
+
+/* clang-format off */
+enum { nvic_iser = 0, nvic_icer = 32, nvic_ispr = 64, nvic_icpr = 96, nvic_iabr = 128,
+	nvic_ip = 192 };
+/* clang-format on */
+
+
+struct {
+	intr_handler_t irqs[SIZE_INTERRUPTS];
+	volatile u32 *nvic;
+} irq_common;
+
+
+static void interrupts_nvicSetIRQ(s8 irqn, u8 state)
+{
+	volatile u32 *ptr = irq_common.nvic + ((u8)irqn >> 5) + (state ? nvic_iser : nvic_icer);
+	*ptr = 1 << (irqn & 0x1F);
+
+	hal_cpuDataSyncBarrier();
+	hal_cpuInstrBarrier();
+}
+
+
+static void interrupts_nvicSetPriority(s8 irqn, u32 priority)
+{
+	volatile u32 *ptr;
+
+	ptr = ((u32 *)(irq_common.nvic + nvic_ip)) + (irqn / 4);
+
+	*ptr = (priority << (8 * (irqn % 4)));
```

Comment:
Undefined behavior with negative shift exponent if `irqn < 0`, is that correct? Maybe `irqn` should be of type `u8` instead of `s8`.

Reply:
Yeah, `u8` seems to be much better.